### PR TITLE
Support for .ready verification event (MSC2366) & other things

### PR DESCRIPTION
**Pull request by bwindels**
MSC: https://github.com/uhoreg/matrix-doc/blob/key_verification_accept/proposals/2366-key-verification-accept.md
Part of https://github.com/vector-im/riot-web/issues/11224
Required by https://github.com/matrix-org/matrix-react-sdk/pull/3796

So https://github.com/matrix-org/matrix-js-sdk/pull/1109 formalized the request object that was already used in `Crypto/index`. This PR further puts all of the responsibilities of `KeyVerificationStateObserver` (react-sdk) into `VerificationRequest` (js-sdk) so the `react-sdk` can use the latter directly to drive the UI in https://github.com/matrix-org/matrix-react-sdk/pull/3796. This makes it easier to keep all the parts that need to response to the state of a request in sync (toast, event tiles and right panel).

Most of the work in this PR went into displaying historical request (e.g. loaded from cache, or when back-paginating) correctly.

In order to make this work, I had to rewrite the state machine logic in `VerificationRequest`, and ended up storing all events by sender and type to not rely on order, because when loading events from cache the events are be emitted in reverse chronological order (e.g. receiving the .done event first and .request event last). To prevent the verifier or the verification request sending out any events (like .cancel) while replaying the events of a historical request, an `observeOnly` flag was added. This flag is turned off when receiving a non-live event (changes had to be made to determine this more reliably, e.g. when loading from cache), or when an event is too old, or when we detect the request is between two other parties (e.g. the user in the DM and another one of my own devices).

In-room requests are now also stored per roomId, not userId, as determining the other party of the request can be hard when the .request is not the first event received (when loading from cache). We need to take determine the other part for that event and ignore any other senders after that. For this reason the request lookup was abstracted into `InRoomRequests` and `ToDeviceRequests`.

The state machine now also relies on remote echos being passed to `handleEvent` (which are faked for to_device messages) as this handles replaying historical events, sending, and handling live events in one code path. Not relying on local echo for verification actions anymore also mean I could delete `RequestCallbackChannel`, which was used to inform the `VerificationRequest` that the verifier was sending something and to do local echo of that event.

As said, this PR also adds support for the .ready event as specified in MSC2366. This turns accepting a request into a separate step than choosing a verification method. To support this, `VerificationRequest` gained an `accept` method and `waitForVerifier` had to be changed as getting a verifier (when receiving a `.start`) now requires calling `accept()` first and not just waiting.

`VerificationRequest` now also handles cancelling with a timeout before the request is accepted (verifier has its own timeout after accepting a request).

#### Breaking change
https://github.com/matrix-org/matrix-js-sdk/pull/1140/commits/2da725340c3950d2e00598db99991131bb2405d7 changes the return type of `client.acceptVerification` and `client.acceptVerificationDM` to return a promise of `VerificationRequest` object instead of a promise of the verifier. Also `client.acceptVerificationDM` was deleted, requests should now be accepted through `VerificationRequest.accept`. See commit message why this was needed. This being a breaking change, it might warrant bumping the major version number.

#### Cleanups still to do:
 - [x] fix lint
 - [x] fix test failure
 - [ ] remove debug logging

The commit log is a bit of a mess because of having to rewrite the state machine and at the same time adding .ready support. I had planned to restructure the commits but ran out of time, apologies to the reviewer for this.

### CODE REVIEW

**Comment by bwindels** [comment]
Meant this to be a draft PR, still need to improve comments.

File: src/crypto/index.js [hunk 1]
```diff
@@ -2471,76 +2446,80 @@ Crypto.prototype._onKeyVerificationMessage = function(event) {
...
-    const sender = event.getSender();
-    let requestsByTxnId = requestsMap.get(sender);
+    let request = requestsMap.getRequest(event);
     let isNewRequest = false;
-    let request = requestsByTxnId && requestsByTxnId.get(transactionId);
     if (!request) {
         request = createRequest(event);
         // a request could not be made from this event, so ignore event
         if (!request) {
+            console.log(`Crypto: could not find VerificationRequest for ${event.getType()}, and could not create one, so ignoring.`);
```
Comment (turt2live) [inline src/crypto/index.js]: in the js-sdk we use `logger` instead of console.

**Comment by turt2live** [comment]
I've merged in develop after I completely refactored the entire build system.

File: src/client.js [hunk 2]
```diff
@@ -894,8 +894,8 @@ MatrixClient.prototype.acceptVerificationDM = function(event, method) {
  * @param {Array} devices array of device IDs to send requests to.  Defaults to
  *    all devices owned by the user
  *
- * @returns {Promise<module:crypto/verification/Base>} resolves to a verifier
- *    when the request is accepted by the other user
+ * @returns {Promise<module:crypto/verification/request/VerificationRequest>} resolves to a VerificationRequest
```
[the remark this answers is not in the document]
Comment (bwindels) [reply src/client.js]: Indeed, see the breaking change comment in the description.

**Pull request merged**